### PR TITLE
8286972: Support the new loop induction variable related PopulateIndex IR node on x86

### PR DESCRIPTION
**Pull request by sviswa7**
This PR adds x86 backend support for the new loop induction variable related PopulateIndex IR node. 
This IR node was added as part of [JDK-8280510](https://bugs.openjdk.java.net/browse/JDK-8280510).

The performance numbers are as follows:
Before:
Benchmark                   (count)   Mode  Cnt       Score       Error  Units
IndexVector.exprWithIndex1    65536  thrpt    3   64556.552 ±  1126.396  ops/s
IndexVector.exprWithIndex2    65536  thrpt    3   22117.050 ± 11452.098  ops/s
IndexVector.indexArrayFill    65536  thrpt    3  117776.383 ±  1120.957  ops/s

After:
Benchmark                   (count)   Mode  Cnt       Score       Error  Units
IndexVector.exprWithIndex1    65536  thrpt    3  203180.290 ±  2147.807  ops/s
IndexVector.exprWithIndex2    65536  thrpt    3  274132.756 ±  6853.393  ops/s
IndexVector.indexArrayFill    65536  thrpt    3  374165.202 ± 46930.779  ops/s

Please review.

Best Regards,
Sandhya

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8286972](https://bugs.openjdk.java.net/browse/JDK-8286972): Support the new loop induction variable related PopulateIndex IR node on x86

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [727daec0](https://git.openjdk.java.net/jdk/pull/8778/files/727daec095e9779724fc5924dbf41d8c8de7c98f)
 * [Jatin Bhateja](https://openjdk.java.net/census#jbhateja) (@jatin-bhateja - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8778/head:pull/8778` \
`$ git checkout pull/8778`

Update a local copy of the PR: \
`$ git checkout pull/8778` \
`$ git pull https://git.openjdk.java.net/jdk pull/8778/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8778`

View PR using the GUI difftool: \
`$ git pr show -t 8778`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8778.diff">https://git.openjdk.java.net/jdk/pull/8778.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (vnkozlov) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2285]: This whole block could be under `#ifdef ASSERT`.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (vnkozlov) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2282]: What are acceptable values of `vlen_enc`?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (vnkozlov) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2314]: The `} else {` case will be also executed on on KNL CPU. Did you tested with `-XX:+UseKNLSetting`?

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1475]: Please, don't do `fallthrough` - `RoundVF` is not related to `PopulateIndex`. Why not `if (!is_LP64 || UseAVX < 2)`?
Are there limitations in 32 bits or you don't want spend time on not major platform (which is also understandable)?

File: src/hotspot/cpu/x86/x86.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1821]: Use `{}` according to our style.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2282]: For KNL, PopulateIndex support is limited to 256-bit as we need avx512bw() for the 512-bit support.
For other AVX2 and AVX512 architectures, all vector widths up to and including 512-bit are supported.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:1475]: It is not a major platform so I didn't spend time on it.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2316]: Yes, this part will be executed on KNL CPU. 
I did run the compiler tests with UseKNLSetting and didn't see any issue.

**Comment by sviswa7** [comment]
@vnkozlov I will look into adding the IR framework and regression test.

**Review by vnkozlov — COMMENTED** [review]
Tier1-4 testing passed.

**Comment by sviswa7** [comment]
@vnkozlov thanks a lot.

**Comment by sviswa7** [comment]
@vnkozlov I have added the IR framework jtreg test. Please review.

**Review by vnkozlov — APPROVED** [review]
Good.

You need second review

**Comment by sviswa7** [comment]
@vnkozlov Thanks a lot for the review.

File: test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java
Comment (pfustc) [inline test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java:29]: Can we simplify `os.arch=="amd64" | os.arch=="x86_64"` to `os.simpleArch == "x64"` ?

This test runs on x86 only. It would be nice if it can run on AArch64 as well. So perhaps something like
```
 28 * @requires (os.simpleArch == "x64" & vm.cpu.features ~= ".*avx2.*") |
 29 *           (os.simpleArch == "aarch64" & vm.cpu.features ~= ".*sve.*")
```

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2311]: Can't we use single and double precision broadcasts for floating point types, like you have done in else part
It may save domain switch over penalty (Section 3.5.2.2 Bypass between Execution Domains, Intel® 64 and IA-32 Architectures Optimization Reference Manual)

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86.ad:8272]: Ideally assertion should be the first statement in a block, since they determine the pre-conditions under which code should executed.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86.ad:8269]: Matcher::vector_length can be directly used instead of following computation in line 8274
    vlen_in_bytes/type2aelembytes(elem_bt)

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86.ad:8288]: Same as above.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86.ad:8291]: Same as above

File: test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java:26]: PR id missing.

File: test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java:39]: Small array size around 10K may work, we can also tune CompileThresholdScaling.

File: test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java:71]: post-incrementation for consistency.

File: test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java
Comment (jatin-bhateja) [inline test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java:107]: post-increment induction.

File: test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java
Comment (sviswa7) [reply test/hotspot/jtreg/compiler/vectorization/TestPopulateIndex.java]: @pfustc  Yes, changed the requires to as suggested by you.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (sviswa7) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2311]: The floating point broadcast doesn't take the gpr as second source.

**Comment by sviswa7** [comment]
@jatin-bhateja Thanks a lot for the review. Your review comments are implemented.

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:2311]: A prior move as in else part may be emitted for consistency or you want to keep floating point broadcasts only for else part.

**Comment by jatin-bhateja** [comment]
LGTM. Thanks

**Review by jatin-bhateja — APPROVED** [review]

**Comment by vnkozlov** [comment]
I have to re-test it since code change in .ad file.

**Comment by sviswa7** [comment]
@vnlozlov Thanks a lot, I will wait for your go ahead.

**Comment by vnkozlov** [comment]
My testing passed. You are good to push.

File: src/hotspot/cpu/x86/x86.ad
Comment (fg1417) [inline src/hotspot/cpu/x86/x86.ad:8270]: May I ask why use `Matcher::vector_length()` here, rather than `Matcher::vector_length_in_bytes()`, for `load_iota_indices()`? Thanks.

**Pull request closed without merging**
